File: apps/mobile/src/features/product-scanner/product-match.ts
import type { ProductScanResult } from "@/features/product-scanner/types";
import type { MobileProduct } from "@/lib/inventory-types";

const normalized = (value: string | null | undefined) =>
  value
    ?.normalize("NFKD")
    .toLocaleLowerCase()
    .replace(/[^\p{L}\p{N}]+/gu, " ")
    .trim() ?? "";

const words = (value: string) => new Set(value.split(/\s+/).filter((part) => part.length > 1));

const overlap = (left: string, right: string) => {
  const leftWords = words(left);
  const rightWords = words(right);
  if (leftWords.size === 0 || rightWords.size === 0) return 0;
  let shared = 0;
  for (const word of leftWords) if (rightWords.has(word)) shared += 1;
  return shared / Math.max(leftWords.size, rightWords.size);
};

const matchScore = (product: MobileProduct, result: ProductScanResult, recognizedText: string) => {
  const productName = normalized(product.name);
  const inferredName = normalized(result.name);
  const ocr = normalized(recognizedText);
  let score = 0;

  if (inferredName && inferredName === productName) score = 0.68;
  else if (productName.length >= 3 && ocr.includes(productName)) score = 0.62;
  else if (inferredName) score = overlap(productName, inferredName) * 0.64;

  const inferredComposition = normalized(result.composition);
  const productComposition = normalized(product.composition);
  if (inferredComposition && productComposition) {
    const compositionScore =
      inferredComposition === productComposition
        ? 1
        : overlap(productComposition, inferredComposition);
    score += compositionScore * 0.18;
    if (compositionScore < 0.25) score -= 0.28;
  }

  const inferredStrength = normalized(result.strength);
  const productStrength = normalized(product.strength);
  if (inferredStrength && productStrength)
    score += inferredStrength === productStrength ? 0.24 : -0.55;
  return Math.min(1, Math.max(0, score));
};

export const findProductMatch = (
  products: ReadonlyArray<MobileProduct>,
  result: ProductScanResult,
  recognizedText: string,
) => {
  const ranked = products
    .map((product) => ({ product, score: matchScore(product, result, recognizedText) }))
    .sort((left, right) => right.score - left.score);
  const best = ranked[0];
  const runnerUp = ranked[1];
  if (!best || best.score < 0.5) return null;
  if (runnerUp && best.score - runnerUp.score < 0.08) return null;
  return best.product;
};

File: apps/android/core/src/main/kotlin/com/tabaaq/mobile/core/scan/ProductMatch.kt
package com.tabaaq.mobile.core.scan

import com.tabaaq.mobile.core.catalog.CatalogProduct
import java.text.Normalizer

object ProductMatch {
    fun find(
        products: List<CatalogProduct>,
        result: ProductScanResult,
        recognizedText: String,
    ): CatalogProduct? {
        val ranked =
            products
                .map { it to score(it, result, recognizedText) }
                .sortedByDescending { it.second }
        val best = ranked.firstOrNull() ?: return null
        val runnerUp = ranked.getOrNull(1)
        if (best.second < 0.5) return null
        if (runnerUp != null && best.second - runnerUp.second < 0.08) return null
        return best.first
    }

    internal fun score(
        product: CatalogProduct,
        result: ProductScanResult,
        recognizedText: String,
    ): Double {
        val productName = normalized(product.name)
        val inferredName = normalized(result.name)
        val ocr = normalized(recognizedText)
        var score = 0.0
        score +=
            when {
                inferredName.isNotEmpty() && inferredName == productName -> 0.68
                productName.length >= 3 && ocr.contains(productName) -> 0.62
                inferredName.isNotEmpty() -> overlap(productName, inferredName) * 0.64
                else -> 0.0
            }
        val inferredComposition = normalized(result.composition)
        val productComposition = normalized(product.composition)
        if (inferredComposition.isNotEmpty() && productComposition.isNotEmpty()) {
            val compositionScore =
                if (inferredComposition == productComposition) 1.0 else overlap(productComposition, inferredComposition)
            score += compositionScore * 0.18
            if (compositionScore < 0.25) score -= 0.28
        }
        val inferredStrength = normalized(result.strength)
        val productStrength = normalized(product.strength)
        if (inferredStrength.isNotEmpty() && productStrength.isNotEmpty()) {
            score += if (inferredStrength == productStrength) 0.24 else -0.55
        }
        if (result.unitsPerPack != null) {
            score += if (result.unitsPerPack == product.unitsPerPack) 0.22 else -0.7
        }
        return score.coerceIn(0.0, 1.0)
    }

    private fun normalized(value: String?): String {
        if (value.isNullOrBlank()) return ""
        val decomposed = Normalizer.normalize(value, Normalizer.Form.NFKD).lowercase()
        return decomposed.replace(Regex("""[^\p{L}\p{N}]+"""), " ").trim()
    }

    private fun words(value: String): Set<String> = value.split(Regex("""\s+""")).filter { it.length > 1 }.toSet()

    private fun overlap(
        left: String,
        right: String,
    ): Double {
        val leftWords = words(left)
        val rightWords = words(right)
        if (leftWords.isEmpty() || rightWords.isEmpty()) return 0.0
        val shared = leftWords.count { it in rightWords }
        return shared.toDouble() / maxOf(leftWords.size, rightWords.size)
    }
}
